sdk/src/main/scala/coop/rchain/sdk/dag/merging/data/MergeableScope.scala

```diff
@@ -4,57 +4,57 @@ import coop.rchain.sdk.dag.merging.data.MergeableScope.FinalizedDeploys
-    (finalityState.keySet -- finalityFringe -- finalityFringe.flatMap(_.seen)).isEmpty,
-    "Invalid MergeableScope"
+    (latestFringe.flatMap(_.seen) -- finalizationIndex.keySet).isEmpty,
+    "Index is unavailable for some finalized messages."
   )
 
-  // All messages that are finalized. Can be empty.
-  lazy val finalScope: Set[Block[B, D]] = finalityState.keySet
+  // Final fringe + all messages seen from the fringe
+  lazy val finalScope: Set[B] = finalizationIndex.keySet
 
   // All messages that are not finalized.
-  lazy val conflictScope = latestMsgs ++ latestMsgs.flatMap(_.seen) -- finalScope
-
-  // Fringes required for the state to be complete
-  lazy val allFringes: Set[Set[B]] = finalityFringe + finalityFringe.map(_.fringe)
-
-  // State is complete when fringe for each message in the conflict set is present in the state
-  lazy val isComplete = (conflictScope.map(_.fringe) -- allFringes).isEmpty
+  lazy val conflictScope =
+    (latestMsgs ++ latestMsgs.flatMap(_.seen) --
+      latestFringe -- latestFringe.flatMap(_.seen))
+      .map(mb)
 
   // All deploys which are not finalized.
   lazy val conflictSet: Set[D] = conflictScope.flatMap(_.deploys)
 
   // Deploys finalized
-  lazy val rejectedFinally: Set[D] = finalityState.flatMap(_._2.rejected).toSet
-  lazy val acceptedFinally: Set[D] = finalityState.flatMap(_._2.accepted).toSet
+  lazy val rejectedFinally: Set[D] = finalizationIndex.flatMap(_._2.rejected).toSet
+  lazy val acceptedFinally: Set[D] = finalizationIndex.flatMap(_._2.accepted).toSet
 }
 
 object MergeableScope {
 
+  /** Part of block index - deploys rejected and accepted on finalization. */
   final case class FinalizedDeploys[D](rejected: Set[D], accepted: Set[D])
 
-  /**
-    * Incomplete view tells that node is still in bootstrap mode and unable to validate blocks on its own.
-    */
-  final case class Incomplete()
+  /** MergeableScope is incomplete because some blocks do not have indices available. */
+  final case class IncompleteMergeableScope[B](indicesMissing: Set[B])
+
+  /** State is complete when for each message in the conflict scope the final fringe is available. */
+  def checkComplete[B, D](s: MergeableScope[B, D]): Set[B] =
+    s.conflictScope.flatMap(_.fringe) -- s.finalScope
 
-  // Attempt to create Casper snapshot.
-  // Returns either the snapshot that can be used for validation / block creation or Incomplete,
-  // which means node is in bootstrap mode.
   def apply[B, D](
       latestMsgs: Set[B],
-      msgSet: Map[B, Block[B, D]],
-      fringeSet: Map[Set[B], FinalizedDeploys[D]],
-      finalizerOutput: Option[Set[B]]
-  ): Either[Incomplete, MergeableScope[B, D]] = {
-    val v = new MergeableScope[B, D](latestMsgs, finalizerOutput, fringeSet)(msgSet)
-    if (v.isComplete) v.asRight[Incomplete]
-    else Incomplete().asLeft[MergeableScope[B, D]]
+      fringe: Set[B],
+      finalizationIndex: Map[B, FinalizedDeploys[D]],
+      msgSet: Map[B, MergeableBlock[B, D]]
```

Comment:
Instead of duplicating the message map we should have another map with data not available in existing message map like deploys e.g. deploys Map[MId, Set[<deploy_ids>]].